flexget/plugins/filter/upgrade.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import unicode_literals, division, absolute_import
+from builtins import *
+from collections import defaultdict
+from datetime import datetime
+from sqlalchemy import Column, String, Unicode, DateTime, Boolean
+import logging
+from os.path import expanduser
+
+from flexget import db_schema, plugin
+from flexget.entry import Entry
+from flexget.utils.database import quality_property
+from flexget.db_schema import Session
+from flexget.event import event
+
+log = logging.getLogger('upgrade')
+
+Base = db_schema.versioned_base('upgrade', 0)
+
+
+class EntryUpgrade(Base):
+    __tablename__ = 'upgrade'
+
+    id = Column(Unicode, primary_key=True, index=True)
+    title = Column(Unicode)
+    _quality = Column('quality', String)
+    quality = quality_property('_quality')
+    proper = Column(Boolean)
+    added = Column(DateTime, index=True)
+
+    def __init__(self):
+        self.added = datetime.now()
+
+    def __str__(self):
+        return '<Upgrade(id=%s,added=%s,quality=%s)>' % \
+               (self.id, self.added, self.quality)
+
+
+def reject_lower(self, quality, entries):
+    for entry in entries:
+        if quality > entry['quality']:
+            entry.reject('Higher quality already downloaded')
+
+
+def group_entries(entries, identified_by):
+    grouped_entries = defaultdict(list)
+
+    # Group by Identifier
+    for entry in entries:
+        identifier = entry.get('id') if identified_by == 'auto' else entry.render(identified_by)
+        if identifier:
+            grouped_entries[identifier.lower()].append(entry)
+        else:
+            log.debug('No identifier found for', entry['title'])
+
+    return grouped_entries
+
+
+class LazyUpgrade(object):
+    schema = {
+        'oneOf': [
+            {'type': 'boolean'},
+            {
+                'type': 'object',
+                'properties': {
+                    'identified_by': {'type': 'string', 'default': 'auto'},
+                    'tracking': {'type': 'boolean', 'default': True},
+                    'on_lower': {'type': 'string', 'enum': ['accept', 'reject', 'fail', 'skip'], 'default': 'skip'}
+                },
+                'additionalProperties': False
+            }
+        ]
+    }
+
+    def prepare_config(self, config):
+        if not config or isinstance(config, bool):
+            return {'identified_by': 'auto', 'on_lower': 'skip', 'tracking': True}
```

Comment:
I'd rather the defaults were set _only_ in `prepare_config` with `dict.setdefault()`

Reply:
I wish there was a better way.. I wish we didn't near prepare_config at all. But we need it because it's oneOf

Reply:
I don't mind prepare_config.